torch/csrc/jit/pybind_utils.h

```diff
@@ -88,82 +97,146 @@ inline c10::optional<TypePtr> unifyOrInitializeType(
+
+    for (py::handle elem : tuple) {
+      auto type_match = tryToInferType(elem);
+      if (type_match.type) {
+        element_types.push_back(*type_match.type);
+      } else {
+        // Forward error message along
+        return type_match.errMsg;
+      }
     }
-    return TypedIValue(Tuple::create(s), TupleType::create(t));
+    return MatchTypeReturn(TupleType::create(element_types));
   } else if (PyDict_Check(input.ptr())) {
     // Check to make sure we can generate useful input/output types
     auto dict = py::cast<py::dict>(input);
-    c10::impl::GenericDictPtr elems = c10::impl::make_generic_dict();
-
     size_t len = py::len(dict);
     if (!len) {
-      AT_ERROR("Dictionary inputs must have entries.");
+      return MatchTypeReturn("Dictionary inputs must have entries");
     }
-    elems.reserve(len);
 
-    TypePtr keyType = nullptr;
-    TypePtr valueType = nullptr;
+    TypePtr key_type = nullptr;
+    TypePtr value_type = nullptr;
+
     for (auto entry : dict) {
-      auto keyInfo = toDictKeyIValue(entry.first);
-      auto valInfo = toTypedIValue(entry.second);
-      auto unifiedKey = unifyOrInitializeType(keyType, keyInfo.second);
-      auto unifiedValue = unifyOrInitializeType(valueType, valInfo.second);
-      if (!unifiedKey || !unifiedValue) {
-        AT_ERROR(
+      // Try to infer the key type and unify it with the existing one
+      auto entry_key_type_match = tryToInferType(entry.first);
+      if (!entry_key_type_match.type) {
+        return entry_key_type_match.errMsg;
+      }
+      auto unified_key =
+          unifyOrInitializeType(key_type, *entry_key_type_match.type);
+      if (!unified_key) {
+        return MatchTypeReturn(
             "Dictionary inputs to traced functions must have consistent type");
       }
-      keyType = *unifiedKey;
-      valueType = *unifiedValue;
-      elems.insert(keyInfo.first, valInfo.first);
+
+      // Try to infer the value type and unify it with the existing one
+      auto entry_value_type_match = tryToInferType(entry.second);
+      if (!entry_value_type_match.type) {
+        return entry_value_type_match.errMsg;
+      }
+      auto unified_value =
+          unifyOrInitializeType(value_type, *entry_value_type_match.type);
+      if (!unified_value) {
+        return MatchTypeReturn(
+            "Dictionary inputs to traced functions must have consistent type");
+      }
+
+      key_type = *unified_key;
+      value_type = *unified_value;
     }
-    return TypedIValue(
-        at::ivalue::GenericDict::create(std::move(elems)),
-        DictType::create(keyType, valueType));
+    return MatchTypeReturn(DictType::create(key_type, value_type));
   } else if (PyList_Check(input.ptr())) {
     auto list = py::cast<py::list>(input);
-    std::vector<IValue> elems;
     size_t len = py::len(list);
     if (!len) {
-      AT_ERROR("List trace inputs must have elements");
+      return MatchTypeReturn("List trace inputs must have elements");
```

Comment:
Trace is too specific of an error message now

Reply:
What do you mean? This is only for lists

Reply:
It says `List trace inputs`